exercises/sat/descr.md

```diff
@@ -0,0 +1,616 @@
+
+*Hint.* Several implementations are possible. We suggest maintaining an
+infinite array of Booleans, so as to allow `mem`, `add` and `remove` to
+operate in amortized constant time. Some thought is then required to implement
+`pick`. There are several ways of doing so; this is up to you!
+
+## Implementing a Simple SAT Solver
+
+At last, we come to the heart of the matter,
+that is, implementing a SAT solver.
+
+We have written some code for you,
+which sets up four data structures:
+
+* A set `Undecided` of undecided variables.
+
+* An array `C` of all clauses.
+
+* An undo trail `Trail`.
+
+* An array `value` that maps every decided variable to its Boolean value.
+  (The Boolean value associated with an undecided variable is irrelevant.)
+
+The array `C.clauses` has type `clause option InfiniteArray.t`. The content of
+this array changes over time: indeed, a clause is simplified when a variable
+that occurs in it becomes decided. The empty clause, which is unsatisfiable,
+is represented by `Some []`. A clause that has been satisfied and therefore
+has disappeared is represented by `None`.
+
+The purpose of the undo trail is to help us undo changes to the other data
+structures, namely `Undecided` and `C`. (It turns out that it is never
+necessary to undo a modification of the array `value`.) This is apparent
+in the next two questions.
+
+**Question D.** (Ungraded.) Implement the function `mark_decided` so that
+`mark_decided x` takes the variable `x` out of the set `Undecided` and pushes
+a suitable undo action onto the trail.
+
+**Question E.** (Ungraded.) Implement the function `set_clause` so that
+`set_clause i clause` writes `clause` at index `i` in the array `C.clauses`
+and pushes a suitable undo action onto the trail.
+
+As mentioned above, when a variable `x` becomes decided, that is, when the
+value of `x` is set to `false` or `true`, we can simplify each of the clauses
+where `x` occurs. Every clause where the positive literal `x` appears is now
+satisfied and disappears. Every clause where the negative literal `~x` appears
+can be simplified by removing this literal. If the clause becomes empty, then
+a contradiction has been reached and the exception `UNSAT` must be raised.
+
+**Question F.** (Ungraded.) Implement the function `set_value` so that
+`set_value x p` sets the value of `x` to `p`, simplifies all of the
+clauses where the variable `x` occurs, and raises `UNSAT` if one
+such clause becomes empty.
+
+The reason why the Boolean satisfiability problem is hard is that a huge tree
+of candidate solutions must be explored. When a variable `x` is undecided, we
+usually have no idea whether `x` should be set to `false` or to `true`, so we
+have to explore both branches.
+
+There *is* a situation, however, where it is easy to determine what the
+appropriate value of `x` should be. This situation arises when the variable
+`x` occurs in a unit clause, that is, a clause that contains just one literal.
```

Comment:
This definition may come a bit too late.

Reply:
I have added a forward pointer from the previous mention of "unit clauses" to this place.